fbx_document.cpp

```diff
@@ -1821,492 +1513,354 @@ Vector<Transform3D> FBXDocument::_decode_accessor_as_xform(Ref<FBXState> p_state
-				const Vector<Vector3> &vertices = array[Mesh::ARRAY_VERTEX];
-				ERR_FAIL_COND_V(vertices.size() == 0, ERR_PARSE_ERROR);
-				Vector<int> indices;
-				const int vs = vertices.size();
-				indices.resize(vs);
-				{
-					int *w = indices.ptrw();
-					for (int k = 0; k < vs; k += 3) {
-						w[k] = k;
-						w[k + 1] = k + 2;
-						w[k + 2] = k + 1;
-					}
-				}
-				array[Mesh::ARRAY_INDEX] = indices;
-			}
+					bones.resize(vertex_num * num_skin_weights);
+					weights.resize(vertex_num * num_skin_weights);
+					for (int32_t vertex_i = 0; vertex_i < vertex_num; vertex_i++) {
+						uint32_t fbx_vertex_index = fbx_mesh->vertex_indices[indices[vertex_i]];
+						ufbx_skin_vertex skin_vertex = fbx_skin->vertices[fbx_vertex_index];
 
-			bool generate_tangents = (primitive == Mesh::PRIMITIVE_TRIANGLES && !a.has("TANGENT") && a.has("TEXCOORD_0") && a.has("NORMAL"));
+						float total_weight = 0.0f;
+						for (int32_t i = 0; i < int32_t(skin_vertex.num_weights); i++) {
+							ufbx_skin_weight skin_weight = fbx_skin->weights[skin_vertex.weight_begin + i];
+							bones.write[vertex_i * num_skin_weights + i] = int(skin_weight.cluster_index);
+							weights.write[vertex_i * num_skin_weights + i] = float(skin_weight.weight);
+							total_weight += float(skin_weight.weight);
+						}
+						if (total_weight > 0.0f) {
+							for (int32_t i = 0; i < int32_t(skin_vertex.num_weights); i++) {
+								weights.write[vertex_i * num_skin_weights + i] /= total_weight;
+							}
+						}
 
-			Ref<SurfaceTool> mesh_surface_tool;
-			mesh_surface_tool.instantiate();
-			mesh_surface_tool->create_from_triangle_arrays(array);
-			if (a.has("JOINTS_0") && a.has("JOINTS_1")) {
-				mesh_surface_tool->set_skin_weight_count(SurfaceTool::SKIN_8_WEIGHTS);
-			}
-			mesh_surface_tool->index();
-			if (generate_tangents) {
-				//must generate mikktspace tangents.. ergh..
-				mesh_surface_tool->generate_tangents();
-			}
-			array = mesh_surface_tool->commit_to_arrays();
-
-			Array morphs;
-			//blend shapes
-			if (p.has("targets")) {
-				print_verbose("glTF: Mesh has targets");
-				const Array &targets = p["targets"];
-
-				//ideally BLEND_SHAPE_MODE_RELATIVE since gltf2 stores in displacement
-				//but it could require a larger refactor?
-				import_mesh->set_blend_shape_mode(Mesh::BLEND_SHAPE_MODE_NORMALIZED);
-
-				if (j == 0) {
-					const Array &target_names = extras.has("targetNames") ? (Array)extras["targetNames"] : Array();
-					for (int k = 0; k < targets.size(); k++) {
-						String bs_name;
-						if (k < target_names.size() && ((String)target_names[k]).size() != 0) {
-							bs_name = (String)target_names[k];
-						} else {
-							bs_name = String("morph_") + itos(k);
+						// Pad the rest with empty weights
+						for (int32_t i = int32_t(skin_vertex.num_weights); i < num_skin_weights; i++) {
+							bones.write[vertex_i * num_skin_weights + i] = 0; // TODO: What should this be padded with?
```

Comment:
It becomes an DDC problem as any bone is possible. We solved in blender gltf importer by creating a bone at root that weights all the unweighted vertices.

Reply:
One option is to also duplicate the last bone index with zero weight.

Reply:
I don't think that is different from the first bone index with zero weight?